proposals/0000-decl-invisible-params.rst

```diff
@@ -18,61 +18,156 @@ This adds the syntax to fix that not already proposed.
+At least, we can use standalone kind signatures to illustrate the difference if not fix the problem::
+
+  type Foo_0, Foo_1, Foo_2 ::
+    forall k. Type -> forall (b :: k). Type
+
+versus::
+
+  type Bar_0, Bar_1, Bar_2 ::
+    Type -> forall k (b :: k). Type
+
+(pretending for concision's sake we had multiple name SAKS syntax).
 
-  data F :: forall k. forall (a :: k) -> Type
+Scoping in declaration bodies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  data family F :: forall k. (a :: k) -> Type
+These all work (with SAKSs for just clarity---they can be removed)::
 
-For both the ``data`` and ``data family`` declaration, the order in which the ``k`` parameter is quantified changes, but perhaps that isn't so bad on its own.
-Worse is that for the family definition, the implicitly bound ``k`` is something instances can scrutinize, while the other ``k`` is purely parametric.
+  type Foo :: Type -> Type
+  data Foo k = forall (b :: k). MkFoo
 
-Making matter worse still, the latter forms of each cannot be written with all variables explicitly bound.
-How are we suppose to teach these subtle things made even by no explicit way to highlight their differences!
+  type Foo :: forall k. k -> Type
+  data Foo (a :: k) = forall (b :: k). MkFoo
 
-The solution is ``@``-prefixed "invisible parameter", as they are known, just as is already proposed for constructor patterns, in `Proposal 126`_, and lambdas, in proposal `Proposal 155`_.
-The uniformity between all of these should round out the language according to expectations.
+  type Foo :: forall k. k -> Type
+  type Foo (a :: k) = k
+
+  type C :: Type -> Constraint
+  class C k where
+    type F :: k
+
+  type C :: forall k. k -> Constraint
+  class C (a :: k) where
+    type F :: k
+
+But these don't::
+
+  type Foo :: forall k. Type
+  data Foo = forall (b :: k). MkFoo -- k is unbound
+
+  type Foo :: forall k. Type -> Type
+  data Foo a = forall (b :: k). MkFoo -- k is unbound
+
+  type Foo :: forall k. k -> Type
+  type a = k
+
+  type C :: forall k. k -> Constraint
+  class C a where
+    type F :: k
+
+The problem is while free variables in the LHS / head are implicit bound, free variables in the RHS / body are not.
+The ``forall k`` in the SAKSs corresponds the invisible parameter we'd like to use.
+In the ambiguous cases like the first failing example, where no visible parameter references the invisible parameter, there is in fact no way to make the experiment type-check today.
+
+Arity
+~~~~~
+
+For type synonyms and type families, the kind doesn't yet tell the whole story.
+There is also a notion of arity: the number of arguments a type synonym family must be applied if the use of it is to be allowed.
+
+We have the rough intuition that the number of parameters with patterns is the arity.
+The first problem is that implicitly bound variables in the kind, in floating to the LHS, also count::
+
+  -- arity 1
+  type Foo_0 :: forall (b :: k). Type
```

Comment:
Doesn't this fail the forall-or-nothing check?

Reply:
Indeed!

Reply:
There's a way to evade the check at the cost of making the example more contrived:

```
type Foo_0 :: forall. forall (b :: k). Type
```

Reply:
Actually that's still an explicit forall at the beginning, so I think it would still be disallowed. I threw in a leading `Type ->` for now.

Reply:
Actually, that did not work. But here's something that does:

```
type Foo_0 :: Type -> forall (b :: k). Type
data Foo_0 a
```